solution/solveintents.go

```diff
@@ -0,0 +1,216 @@
+// from the batch.
+// Received are treated as expired because they may have been compressed to
+// Solved Intents.
+//
+// The Solver may return a subset and in different sequence the UserOperations
+// and a matching occurs by the hash value of each UserOperation to the bundle
+// UserOperation.
+package solution
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math/big"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+	"unsafe"
+
+	"github.com/blndgs/model"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/goccy/go-json"
+	"github.com/pkg/errors"
+
+	"github.com/stackup-wallet/stackup-bundler/pkg/modules"
+	"github.com/stackup-wallet/stackup-bundler/pkg/userop"
+)
+
+// userOpHashID is the hash value of the UserOperation
+type opHashID string
+
+// batchOpIndex is the index of the UserOperation in the Bundler batch
+type batchOpIndex int
+
+// batchIntentIndices buffers the mapping of the UserOperation hash value -> the index of the UserOperation in the batch
+type batchIntentIndices map[opHashID]batchOpIndex
+
+type IntentsHandler struct {
+	SolverURL    string
+	SolverClient *http.Client
+}
+
+// Verify structural congruence
+var _ = model.UserOperation(userop.UserOperation{})
+
+func New(solverURL string) *IntentsHandler {
+	const httpClientTimeout = 100 * time.Second
+
+	return &IntentsHandler{
+		SolverURL:    solverURL,
+		SolverClient: &http.Client{Timeout: httpClientTimeout},
+	}
+}
+
+// bufferIntentOps caches the index of the userOp in the received batch and creates the UserOperationExt slice for the
+// Solver with cached Hashes and ProcessingStatus set to `Received`.
+func (ei *IntentsHandler) bufferIntentOps(entrypoint common.Address, chainID *big.Int, batchIndices batchIntentIndices, userOpBatch []*model.UserOperation) model.BodyOfUserOps {
+	body := model.BodyOfUserOps{
+		UserOps:    make([]*model.UserOperation, 0, len(userOpBatch)),
+		UserOpsExt: make([]model.UserOperationExt, 0, len(userOpBatch)),
+	}
+	for idx, op := range userOpBatch {
+		if op.HasIntent() {
+			hashID := op.GetUserOpHash(entrypoint, chainID).String()
+
+			// Don't mutate the original op
+			clonedOp := *op
+			body.UserOps = append(body.UserOps, &clonedOp)
+
+			body.UserOpsExt = append(body.UserOpsExt, model.UserOperationExt{
+				OriginalHashValue: hashID,
+				// Cache hash before it changes
+				ProcessingStatus: model.Received,
+			})
+
+			// Reverse caching
+			batchIndices[opHashID(hashID)] = batchOpIndex(idx)
+		}
+	}
+
+	return body
+}
+
+// SolveIntents returns a BatchHandlerFunc that will send the batch of UserOperations to the Solver
+// and those solved to be sent on chain.
+func (ei *IntentsHandler) SolveIntents() modules.BatchHandlerFunc {
+	return func(ctx *modules.BatchHandlerCtx) error {
+		batchIntentIndices := make(batchIntentIndices)
+
+		// cast the received userOp batch to a slice of model.UserOperation
+		// to be sent to the Solver
+		modelUserOps := *(*[]*model.UserOperation)(unsafe.Pointer(&ctx.Batch))
```

Comment:
`unsafe.Pointer` to cast ctx.Batch to a []*model.UserOperation. it's risky because it bypasses Go type safety.

Reply:
Sure, I moved the compiler safety check for the cast right before it. The alternative of repeatedly creating the model objects for every solver request to copy values will put unnecessary GC pressure and unnecessary overhead.

Reply:
fixed in https://github.com/blndgs/bundler/pull/1/commits/c1f8e4e6661105d8ae1dd8a9fe823c26676f764a